osf_tests/factories.py

```diff
@@ -795,6 +794,97 @@ def _create(cls, target_class, *args, **kwargs):
         instance.save()
         return instance
 
+    @classmethod
+    def create_version(cls, *args, **kwargs):
+        source_preprint = kwargs.pop('preprint')
+        update_task_patcher = mock.patch('website.preprints.tasks.on_preprint_updated.si')
+        update_task_patcher.start()
+
+        if not source_preprint.date_published:
+            raise Exception(f'Unable to create a new version as source preprint is still unpublished')
+        if source_preprint.versioned_guids.first().guid.referent.machine_state == 'pending':
+            raise Exception(f'Unable to create a new version as source preprint is still in `pending` state')
+        if not source_preprint.is_latest_version:
+            raise Exception(f'Unable to create a new version as source preprint must be a latest version')
```

Comment:
I think this check shouldn't happen on the `source_version` but on the latest_version. I will make an extra commit to fix this.

Reply:
I also discovered a bug due in Preprints version creation (not in the factory) which I will try to fix part of it in here too and create another bug ticket to fix everywhere else.

Reply:
Made a small compatible change with some TODO comments for the upcoming refactor: https://github.com/CenterForOpenScience/osf.io/pull/10840/commits/5a842c044f33c822f23f3f07400b916830975faf